### ISSUE TITLE
GMT 6.0.0.rc1: BAT file execution error by double-clicking

### ISSUE THREAD

**Issue by liuzhumei** [GenericMappingTools/gmt#994]
**Description of the problem**
The test.bat file was written as:

```
gmt begin map png
gmt coast -Rg -JH10c -Baf -W1p
gmt end 
pause
```
It ran well and output the png image in DOS terminal.
But when I double-clicked the test.bat file, it came with some puzzled problems:

 ```
E:\>gmt begin map png
begin [ERROR]: Unable to create a workflow directory : C:/Users/username/.gmt
/sessions/gmt6.16120

E:\>gmt coast -Rg -JH10c -Baf -W1p
gmt [ERROR]: Shared GMT module not found: coast

ERROR: No module named coast was found.  This could mean one of four cases:
  1. There actually is no such module; please check your spelling.
  2. You used a modern mode module name while running in GMT classic mode.
  3. Module exists in the GMT supplemental library, but the library could not be found.
  4. Module exists in a GMT custom library, but none was specified via GMT_CUSTOM_LIBS.
Shared libraries must be in standard system paths or set via environmental parameter %%PAT
H%%.

E:\>gmt end
gmt [ERROR]: Not available in classic mode
```

**I tried it on Windows 10 (64bit) and an almost clear system for Windows 7 (64bit) to exclude system permission or environment variable interference. It didn't work either.**

**Comment by PaulWessel** [comment]
Looks like double-clicking your bat file runs it as some other user? It looks for C:/Users/username which presumably is not your account and home directory, and since it fails nothing will work after that.   We expect you to run commands in a command window.

**Comment by liuzhumei** [comment]
Actually, username is my account name. The environment variable HOME is set to be `C:\Users\username`, and the GMT session directory `C:\Users\username\.gmt\sessions` exists.
BTW，I have been using GMT-6.0.0_r20419 for months, it works well on my PC.

**Comment by PaulWessel** [comment]
OK, but seem double-clicking runs differently. Google says something about right-clicking to run as Administrator.

**Comment by joa-quim** [comment]
It runs fine for me. But the issue here is why
```
begin [ERROR]: Unable to create a workflow directory : C:/Users/username/.gmt
/sessions/gmt6.16120
```
Don't you have a ``C:/Users/username/.gmt/sessions`` directory?

**Comment by seisman** [comment]
Is it possible that it's caused by separators (`/` and `\`) in the path?

**Comment by PaulWessel** [comment]
If he is running a very old GMT 6 version then the answer is to upgrade first before looking into this further since there are so many changes.

**Comment by joa-quim** [comment]
No, ``/`` or ``\`` (actually ``\\``) is the same on Windows. And no, she says that it was RC1 that gave the problem and the old ``r20419`` worked fine.

**Comment by seisman** [comment]
I can reproduce the issue on Azure Pipelines CI now.

The command is 
```
gmt begin -Vd && gmt coast -R0/10/0/10 -JM6i -Ba -Ggray -Vd && gmt end -Vd
```

The output is (available from this [link](https://dev.azure.com/GenericMappingTools/GMT/_build/results?buildId=1092)):
```
gmt [DEBUG]: Obtained the ppid from parent: 2860
gmt [DEBUG]: Enter: gmtinit_new_GMT_ctrl
gmt [DEBUG]: GMT->session.SHAREDIR = D:/a/1/s/gmt-install-dir/share
gmt [DEBUG]: GMT->session.HOMEDIR = C:/Users/VssAdministrator
gmt [DEBUG]: GMT->session.USERDIR = C:/Users/VssAdministrator/.gmt [created]
gmt [DEBUG]: GMT->session.CACHEDIR = C:/Users/VssAdministrator/.gmt/cache [created]
gmt [DEBUG]: GMT: 0. Will try to find subdir=postscriptlight stem = PSL_custom_fonts suffix=.txt
gmt [DEBUG]: GMT: 1. gmt_getsharepath trying current dir
gmt [DEBUG]: GMT: 2. gmt_getsharepath trying USERDIR C:/Users/VssAdministrator/.gmt
gmt [DEBUG]: GMT: 3. gmt_getsharepath trying USERDIR subdir C:/Users/VssAdministrator/.gmt/postscriptlight
gmt [DEBUG]: GMT: 4. gmt_getsharepath trying SHAREDIR subdir D:/a/1/s/gmt-install-dir/share/postscriptlight
gmt [DEBUG]: GMT: 5. gmt_getsharepath failed
gmt [DEBUG]: Map distance calculation will be Cartesian
gmt [DEBUG]: Exit:  gmtinit_new_GMT_ctrl
gmt [DEBUG]: Enter: New_PSL_Ctrl
gmt [DEBUG]: Exit:  New_PSL_Ctrl
gmt [DEBUG]: Enter: gmt_manage_workflow
gmt [DEBUG]: Exit : gmt_manage_workflow
gmt [DEBUG]: Enter: PSL_beginsession
gmt [DEBUG]: Exit : PSL_beginsession
gmt [DEBUG]: Enter: PSL_setdefaults
gmt [DEBUG]: Exit : PSL_setdefaults
gmt [DEBUG]: Enter: gmtlib_io_init
gmt [DEBUG]: Exit : gmtlib_io_init
gmt [DEBUG]: Enter: gmt_hash_init
gmt [DEBUG]: Exit:  gmt_hash_init
gmt [DEBUG]: Enter: gmt_hash_init
gmt [DEBUG]: Exit:  gmt_hash_init
gmt [DEBUG]: The PROJ_GEODESIC set to Vincenty
gmt [DEBUG]: Enter: gmt_getdefaults
gmt [DEBUG]: Exit:  gmt_getdefaults
gmt [DEBUG]: Enter: gmtlib_plot_C_format
gmt [DEBUG]: Exit:  gmtlib_plot_C_format
gmt [DEBUG]: Enter: gmtinit_get_history
gmt [DEBUG]: Enter: gmt_hash_init
gmt [DEBUG]: Exit:  gmt_hash_init
gmt [DEBUG]: Failed to release lock (Error <158>: The segment is already unlocked.
)
gmt [DEBUG]: Exit:  gmtinit_get_history
gmt [DEBUG]: GMT_Create_Session initialized GMT structure
gmt [DEBUG]: Shared Library # 0 (core). Path = gmt.dll
gmt [DEBUG]: Loading core GMT shared library: gmt.dll
gmt [DEBUG]: Loading GMT plugins from: D:/a/1/s/gmt-install-dir/bin/gmt_plugins
gmt [DEBUG]: Shared Library # 1 (supplements). Path = D:/a/1/s/gmt-install-dir/bin/gmt_plugins/supplements.dll
begin [INFORMATION]: Creating a workflow directory C:/Users/VssAdministrator/.gmt/sessions/gmt6.2860
begin [ERROR]: Unable to create a workflow directory : C:/Users/VssAdministrator/.gmt/sessions/gmt6.2860
begin (gmtlib_free_tmp_arrays): tried to free unallocated memory
gmt [DEBUG]: Entering GMT_Destroy_Session
gmt [DEBUG]: Failed to release lock (Error <158>: The segment is already unlocked.
)
```
From the debug information, you can see the `.gmt` and `.gmt/cache` directories can be created, which means gmt has the written permission to the user's home directory. But gmt fails to create the  session directory.

Setting HOME to other directories like `D:\a\1\s` works.

**Comment by joa-quim** [comment]
@Paul can you make it print also the error code from the gmt_mkdir function? Maybe it can give us a hint.
(I'm in the train now)

**Comment by seisman** [comment]
It shows a "permission denied" error:
```
begin [INFORMATION]: Creating a workflow directory C:/Users/VssAdministrator/.gmt/sessions/gmt6.5716
gmt_mkdir error: Permission denied
begin [ERROR]: Unable to create a workflow directory : C:/Users/VssAdministrator/.gmt/sessions/gmt6.5716
begin (gmtlib_free_tmp_arrays): tried to free unallocated memory
```

**Comment by PaulWessel** [comment]
Makes sense. As I pointed out earlier, googling on the topic of double-clicking vs terminal-running pointed to the option of right-clicking to select "Run as Administrator".  Perhaps the ownership of the scripts must change in order to do this.

**Comment by joa-quim** [comment]
It does not make much sense to me. Clicking or terminal running are done with the same user permissions. Yes, there are tasks that require administrator permissions and one can "Run as Administrator", but running a batch that results in creating a directory under HOME is hardly one of those tasks.

**Comment by joa-quim** [comment]
Funny, this just happened to me but I've not been able to reproduce it (note the *recreate* instead of *create*)
```
julia> subplot(grid="1x1", limits="0/5/0/5", frame="west", F="s7/7", title="VERY VERY", show=1);plot([0 0; 1 1], c="");subplot(stop=true)
gmt_mkdir (last dir) error: Permission denied
begin [ERROR]: Unable to recreate directory : C:/j/.gmt/sessions/gmt6.19236
```

**Comment by joa-quim** [comment]
Yes, Coverity has lots warnings about time things. Can't verify now because the service has been stalled for almost 2 weeks.

**Comment by cugliming** [comment]
I also encountered the same problem when I ran the script. The script was run on bash on ubuntu on windows.
![TIM图片20190704234209](https://user-images.githubusercontent.com/32346042/60677932-758db080-9eb5-11e9-9125-0eb3619911a6.png)

**Comment by joa-quim** [comment]
Which script? Here we are discussing running batch files but you can't run batch files (.bat) from WSL!

**Comment by cugliming** [comment]
> 
> 
> Which script? Here we are discussing running batch files but you can't run batch files (.bat) from WSL!

It was a shell script ran on WSL.

**Comment by joa-quim** [comment]
Meanwhile I have had several of those _permission denied_ errors from Julia. Coverity has this one that seems to touch the problem. But I have no idea how to solve it.
![Capture](https://user-images.githubusercontent.com/537321/60773910-605a9100-a104-11e9-8bcd-3c4e74822a99.JPG)

**Comment by PaulWessel** [comment]
Useful information at https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use
Basically, no quick fix but suggest removing some checks and instead catch the exceptions that occur when we try to actually do stuff (versus just checking with access).

**Comment by joa-quim** [comment]
Unfortunately I bet that we just move the block
```
	if ((here = getcwd (NULL, 0)) == NULL) {	/* Get the current directory */
		GMT_Report (API, GMT_MSG_NORMAL, "Cannot determine current directory!\n");
		return GMT_RUNTIME_ERROR;
	}
```
to before the ``if (access (dir, F_OK)) {`` the Coverity issue will go away but the real one will persist.

**Comment by PaulWessel** [comment]
I think the idea is to not do the access at all and just deal with any errors from getcwd.

**Comment by joa-quim** [comment]
But, for this case, why delete the dir and recreate immediately after?

**Comment by PaulWessel** [comment]
I wish I could confidently answer that, but not sure. Why not just delete all the files and leave the dir.  Maybe paranoia?

**Comment by PaulWessel** [comment]
Maybe you could test by flipping this function around. Set:
`bool delete_dir = !recreate;
`
Then after removing all the files say
```
if (delete_dir && rmdir (dir)) {	/* Unable to delete the directory */
	perror (dir);
	error = GMT_RUNTIME_ERROR;
}
```

**Comment by joa-quim** [comment]
Yes, I think that instead of
```
	if (rmdir (dir)) {	/* Unable to delete the directory */
		perror (dir);
		error = GMT_RUNTIME_ERROR;
	}
	else if (recreate) {	/* Create an empty directory to replace what we deleted */
		if (gmt_mkdir (dir)) {
			GMT_Report (API, GMT_MSG_NORMAL, "Unable to recreate directory : %s\n", dir);
			error = GMT_RUNTIME_ERROR;
		}
	}
```
we can just have
```
if (!recreate && rmdir (dir)) {	/* Unable to delete the directory */
	perror (dir);
	error = GMT_RUNTIME_ERROR;
}
```

**Comment by PaulWessel** [comment]
Sure.  Please build but test all before/after, then see if Julia is happier.

**Comment by joa-quim** [comment]
The point is that the Julia error is random. It happened when somehow ``subplot`` left some trash in mem between calls. We still have issues with non-full cleaning session at end of module calls.

**Comment by joa-quim** [comment]
All tests in ``exmode`` and ``modern`` PASS

**Comment by PaulWessel** [comment]
OK sounds good, commit etc and after merge I will run my tests.

**Comment by seisman** [comment]
I made some changes to the function `gmt_mkdir()` to debug the issue. What I did are:

1. Add print statements to print the variable `_path` passed to `mkdir()` function 
2. Remove the `return -1` statements, so that even if something goes wrong when creating intermediate directory, the function still continues to create the child directories.

The output is:

```
begin [INFORMATION]: Creating a workflow directory C:/Users/VssAdministrator/.gmt/sessions/gmt6.3924
# full path=C:/Users/VssAdministrator/.gmt/sessions/gmt6.3924
## path=C:
gmt_mkdir (intermediate) error: Permission denied
## path=C:/Users
## path=C:/Users/VssAdministrator
## path=C:/Users/VssAdministrator/.gmt
## path=C:/Users/VssAdministrator/.gmt/sessions
## path=C:/Users/VssAdministrator/.gmt/sessions/gmt6.3924
```

When creating the directory `C:`, it reports the permission denied error. If we ignore the error and continues, `gmt_mkdir()` successfully creates the full path recursively. 

We may have two solutions:
1. It doesn't make sense to create a directory `C:` on Windows. Maybe we should skip `C:`
2. Keeping creating the intermediate directories even if there're some errors, and only `return -1` if we fail to create the last directory.

**Comment by PaulWessel** [comment]
Looking at that code which I got from stackoverflow, it does not check if the intermediate directories exist.  But I agree with @seisman; perhaps we can just skip any leading X: "directories" and move to the next real one.  Maybe replace the for loop with a while loop with a slightly modified start:

```
p =(_path[1] == ':') ?_path + 3 :  _path + 1; /* Skip any X: drive designators */
while (*p) {...

p++;
}

```

**Comment by joa-quim** [comment]
It certainly makes to sense trying to create directories ``x:``. Those are disk partitions, not directories.

**Comment by seisman** [comment]
Not sure if we have the same trouble with paths like `/c/users/xxx`, but at least skip `X:` can fix this issue partially.

**Comment by PaulWessel** [comment]
Right, we can see.  Since you are mssing with this, perhaps you can implement the skipping of the X:?

**Comment by seisman** [comment]
OK. I'll try.

**Comment by seisman** [comment]
@joa-quim #1154 has been merged. I hope it can fix this issue. Could you build the latest 6.0 branch and create a Windows installers, so that @liuzhumei and @cugliming may help test if it works.

**Comment by joa-quim** [comment]
Built and updated in the UALG site but since it appears to be so slow to download from there, here is a DB link

https://www.dropbox.com/s/ncgs0d9h5uqtep5/gmt-6.0.0_2019.07.11-win64.exe?dl=0

**Comment by liuzhumei** [comment]
I tested it on Win10 (64bit)  and it finally worked!  *★,°*:.☆(￣▽￣)/$:*.°★* 。
Thank you all ! @PaulWessel @joa-quim @seisman @cugliming

**Comment by seisman** [comment]
I believe the issue was fixed by #1154. Close the issue now. Feel free to reopen the issue if GMT-6.0.0rc3 or later don't work for you.
